content/advent-2018/llvm-ir-and-go.md

```diff
@@ -0,0 +1,311 @@
+
+<!-- TODO: remove 3. Building a toy compiler in Go section? If time allows, add it :) -->
+<!-- 3. Building a toy compiler in Go -->
+
+1. [Quick primer on LLVM IR](#quick-primer-on-llvm-ir)
+2. [LLVM IR library in pure Go](#llvm-ir-library-in-pure-go)
+3. [Closing notes](#closing-notes)
+4. [Further resources](#further-resources)
+
+## Quick primer on LLVM IR
+
+_(For those already familiar with LLVM IR, feel free to [jump to the next section](#llvm-ir-library-in-pure-go))._
+
+[LLVM IR](https://llvm.org/docs/LangRef.html) is a low-level intermediate representation used by the [LLVM compiler framework](http://llvm.org/). You can think of LLVM IR as a platform-independent assembly language with an infinite number of function local registers.
+
+When developing compilers there are huge benefits with compiling your source language to an intermediate representation (IR)[^1] instead of compiling directly to a target architecture (e.g. x86). As many optimization techniques are general (e.g. dead code elimination, constant propagation), these optimization passes may be performed directly on the IR level and thus shared between all targets[^2].
+
+Compilers are therefore often split into three components, the front-end, middle-end and back-end; each with a specific task that takes IR as input and/or produces IR as output.
+
+* **Front-end**: compiles source language to IR.
+* **Middle-end**: optimizes IR.
+* **Back-end**: compiles IR to machine code.
+
+![LLVM compiler pipeline](/postimages/advent-2018/llvm-ir-and-go/llvm_compiler_pipeline.png)
+
+### Example program in LLVM IR assembly
+
+To get a glimps of what LLVM IR assembly may look like, lets consider the following C program.
+
+```c
+int f(int a, int b) {
+	return a + 2*b;
+}
+
+int main() {
+	return f(10, 20);
+}
+```
+
+Using [Clang](https://clang.llvm.org/)[^3], the above C code compiles to the following LLVM IR assembly.
+
+```llvm
+define i32 @f(i32 %a, i32 %b) {
+; <label>:0
+	%1 = mul i32 2, %b
+	%2 = add i32 %a, %1
+	ret i32 %2
+}
+
+define i32 @main() {
+; <label>:0
+	%1 = call i32 @f(i32 10, i32 20)
+	ret i32 %1
+}
+```
+
+By looking at the LLVM IR assembly above, we may observe a few noteworthy details about LLVM IR, namely:
+
+* LLVM IR is statically typed (i.e. 32-bit integer values are denoted with the `i32` type).
+* Local variables are scoped to each function (i.e. `%1` in the `@main` function is different from `%1` in the `@f` function).
+* Unnamed (temporary) registers are assigned local IDs (e.g. `%1`, `%2`) from an incrementing counter in each function.
+* Each function may use an infinite number of registers (i.e. we are not limited to 32 general purpose registers).
+* Global identifiers (e.g. `@f`) and local identifiers (e.g. `%a`, `%1`) are distinguished by their prefix (`@` and `%`, respectively).
+* Most instructions do what you'd think, `mul` performs multiplication, `add` addition, etc.
```

Comment:
Perhaps useful to add is that comments are prefixed with `;` instead of `//`, like many assembly languages but unlike most modern programming languages (like Go). But this list is already on the long side so perhaps it's a bit too much.

Reply:
I like the suggestion, and think I'll try adding it. I know the list is a bit long, but the notice above is that this entire section is intended to refresh aspects of LLVM IR assembly for those quite new to the language, so I think it's fine.